src/cpp/src/block_manager.hpp

```diff
@@ -67,54 +71,116 @@ class KVCacheBlock {
-        block->increment();
-        m_blocks.erase(hash_block->first);
-        return block;
+        auto hash_and_blocks_for_all_layers = std::min_element(std::begin(m_blocks), std::end(m_blocks), [](const auto& lhs, const auto& rhs) -> bool { return lhs.second[0]->get_timestamp() < rhs.second[0]->get_timestamp(); });
+        auto blocks_for_all_layers = hash_and_blocks_for_all_layers->second;
+        auto timestamp = std::chrono::system_clock::now();
+        for (auto& block_ptr : blocks_for_all_layers) {
+            block_ptr->set_timestamp(timestamp);
+            block_ptr->increment();
+        }
+        m_blocks.erase(hash_and_blocks_for_all_layers->first);
+        return blocks_for_all_layers;
     }
 
+    /**
+     *
+     * @return Number of blocks (per layer) currently in the store.
+     */
     size_t num_blocks() const {
         return m_blocks.size();
     }
 };
 
+class CacheStateDumper;
+
+/**
+ * @brief Maintains a pool of KV cache block descriptors (layered as configured at initialization), freeing or allocating
+ * them as requested.
+ */
 class BlockAllocator {
-    std::list<KVCacheBlock::Ptr> m_free_blocks;
-    ov::genai::Evictor m_evictor;
+    std::vector<std::list<KVCacheBlock::Ptr>> m_free_blocks;
     int m_total_num_blocks;
+    friend class CacheStateDumper;
+    size_t m_num_layers;
     bool m_enable_prefix_caching;
+    ov::genai::BlockHashStore m_block_hash_store;
 public:
-    BlockAllocator(int num_blocks, bool enable_prefix_caching) :
-        m_total_num_blocks(num_blocks), m_enable_prefix_caching(enable_prefix_caching) {
-        for (int block_id = 0; block_id < m_total_num_blocks; ++block_id) {
-            m_free_blocks.push_back(std::make_shared<KVCacheBlock>(block_id));
+    /**
+     * Constructs the BlockAllocator.
+     * @param num_blocks Number of KV cache blocks in the free block pool to be owned by this allocator.
+     * @param enable_prefix_caching Whether prefix caching should be enabled for this allocator.
+     * See also the equivalent parameter in ov::genai::ContinuousBatchingPipeline
+     * @param num_layers The number of separate attention layers with KV caches in the LLM associated with the pipeline.
+     * Blocks returned will be vectors with this size, each vector entry to be associated with a separate layer's KV cache.
+     */
+    BlockAllocator(int num_blocks, bool enable_prefix_caching, size_t num_layers = 0) :
+            m_total_num_blocks(num_blocks), m_num_layers(num_layers), m_enable_prefix_caching(enable_prefix_caching), m_block_hash_store(num_layers) {
+        if (m_num_layers == 0) {
+            m_num_layers = 1;
+        }
+        m_free_blocks.resize(m_num_layers);
+        for (auto& per_layer_block_list : m_free_blocks) {
+            for (int block_id = 0; block_id < m_total_num_blocks; ++block_id) {
+                per_layer_block_list.push_back(std::make_shared<KVCacheBlock>(block_id));
```

Comment:
let's use `.reserve` for vector to avoid reallocation during its filling

Reply:
`per_layer_block_list` is actually an `std::list` here